ecdsa/src/hazmat.rs

```diff
@@ -247,6 +205,93 @@ pub fn bits2field<C: PrimeCurve>(bits: &[u8]) -> Result<FieldBytes<C>> {
     Ok(field_bytes)
 }
 
+/// Sign a prehashed message digest using the provided secret scalar and
+/// ephemeral scalar, returning an ECDSA signature.
+///
+/// Accepts the following arguments:
+///
+/// - `d`: signing key. MUST BE UNIFORMLY RANDOM!!!
+/// - `k`: ephemeral scalar value. MUST BE UNIFORMLY RANDOM!!!
+/// - `z`: message digest to be signed. MUST BE OUTPUT OF A CRYPTOGRAPHICALLY
+///        SECURE DIGEST ALGORITHM!!!
+///
+/// # Returns
+///
+/// ECDSA [`Signature`] and, when possible/desired, a [`RecoveryId`]
+/// which can be used to recover the verifying key for a given signature.
+#[cfg(feature = "arithmetic")]
+#[allow(non_snake_case)]
+pub fn sign_prehashed<C, K>(
+    d: &Scalar<C>,
+    k: K,
+    z: &FieldBytes<C>,
+) -> Result<(Signature<C>, RecoveryId)>
+where
+    C: PrimeCurve + CurveArithmetic,
+    K: AsRef<Scalar<C>> + Invert<Output = CtOption<Scalar<C>>>,
+    SignatureSize<C>: ArrayLength<u8>,
+{
+    // TODO(tarcieri): use `NonZeroScalar<C>` for `k`.
+    if k.as_ref().is_zero().into() {
+        return Err(Error::new());
+    }
+
+    let z = <Scalar<C> as Reduce<C::Uint>>::reduce_bytes(z);
+
+    // Compute scalar inversion of 𝑘
+    let k_inv = Option::<Scalar<C>>::from(k.invert()).ok_or_else(Error::new)?;
+
+    // Compute 𝑹 = 𝑘×𝑮
+    let R = ProjectivePoint::<C>::mul_by_generator(k.as_ref()).to_affine();
+
+    // Lift x-coordinate of 𝑹 (element of base field) into a serialized big
+    // integer, then reduce it into an element of the scalar field
+    let r = Scalar::<C>::reduce_bytes(&R.x());
+    let x_is_reduced = r.to_repr() != R.x();
+
+    // Compute 𝒔 as a signature over 𝒓 and 𝒛.
+    let s = k_inv * (z + (r * d));
```

Comment:
Perhaps we should check for `r=0` or `s=0` here, at the lowest level, so that every high level user is automatically protected?

Reply:
That check is performed on the [next line of code when `Signature::from_scalars` is called](https://github.com/RustCrypto/signatures/pull/731/files/928c75bab14778182dd7aa037359d6e746c0e0b4#diff-0328eac1fc4a00e7acac377a2d141c54660be93361d7e010b9e7fdb18c5ad43eR256), which checks both [r and s are non-zero](https://github.com/RustCrypto/signatures/blob/master/ecdsa/src/lib.rs#L253-L255).

Reply:
I think since this function is public API it should return valid data, regardless of whether that data will be later packed into a `Signature`. Or perhaps it could be packed into a `Signature` right away, and if some implementations need to do additional processing, they'll just transform that `Signature`.

Reply:
I'm confused, you want it to return something other than `Signature`?

Reply:
I'd like to prevent casual users who are poking around `Signature` from accidentally discovering it, which is why it's filed away here instead.